src/WebJobs.Script/SlidingWindow.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Azure.WebJobs.Script
+{
+    public class SlidingWindow<TItem>
+    {
+        private readonly TimeSpan _window;
+        private List<Event> _events = new List<Event>();
+        private object _syncLock = new object();
+
+        public SlidingWindow(TimeSpan window)
+        {
+            _window = window;
+        }
+
+        public IEnumerable<TItem> GetEvents(Func<TItem, bool> filter)
+        {
+            lock (_syncLock)
+            {
+                // remove expired events
+                _events = _events.Where(p => (DateTime.Now - p.TimeStamp) < _window).ToList();
+
+                return _events.Select(p => p.Item).Where(filter);
```

Comment:
Because you're not calling .ToList() on the return type here, its possible that the result of calling GetEvents will change after the call completes (because it will be an IEnumerable that is not evaluated yet). This means _events could be in the middle of enumeration when AddEvent is called and the lock won't prevent a "collection modified during enumeration" error...

Reply:
Good catch - fixed :)